File: snippets/ts/src/use-cases/event-processing/user-feed.ts
import * as restate from "@restatedev/restate-sdk";

class SocialMediaPost {}

function createPost(userId: string, post: SocialMediaPost) {
  return undefined;
}

function getPostStatus(postId: any) {
  return "pending";
}

function updateUserFeed(userId: string, postId: any) {
  return undefined;
}

const PENDING = "pending";

// <start_here>
const userFeed = restate.object({
  name: "userFeed",
  handlers: {
    processPost: async (ctx: restate.ObjectContext, post: SocialMediaPost) => {
      const userId = ctx.key;

      const postId = await ctx.run(() => createPost(userId, post));

      // Wait for processing to complete with durable timers
      while ((await ctx.run(() => getPostStatus(postId))) === PENDING) {
        await ctx.sleep({ seconds: 5 });
      }

      await ctx.run(() => updateUserFeed(userId, postId));
    },
  },
});
// <end_here>

File: snippets/java/src/main/java/usecases/eventprocessing/eventtransactions/UserFeed.java
package usecases.eventprocessing.eventtransactions;

import static usecases.eventprocessing.eventtransactions.utils.Stubs.*;

import dev.restate.sdk.Restate;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.VirtualObject;
import java.time.Duration;
import usecases.eventprocessing.eventtransactions.utils.SocialMediaPost;

// <start_here>
@VirtualObject
public class UserFeed {
  @Handler
  public void processPost(SocialMediaPost post) {
    String userId = Restate.key();

    String postId = Restate.run("create-post", String.class, () -> createPost(userId, post));

    while (Restate.run("get-post-status", String.class, () -> getPostStatus(postId))
        .equals("PENDING")) {
      Restate.sleep(Duration.ofSeconds(5));
    }

    Restate.run("update-user-feed", () -> updateUserFeed(userId, postId));
  }
}
// <end_here>
